broadcast/broadcast.go

```diff
@@ -0,0 +1,307 @@
+// Package broadcast implements a fan-out event stream with overwrite-on-full semantics.
+// Multiple producers can call Send concurrently and multiple consumers can Subscribe.
+// Every consumer receives every message in order unless it has fallen behind far enough
+// that its next sequence has been overwritten; in that case the consumer receives a
+// lag notification (LaggedError) indicating where to resume.
+// Notes:
+//   - Producers are serialised by a global mutex (b.mu), so producer-side throughput
+//     is effectively single-writer (MPMC-safe but not truly parallel on Send).
+//   - The ring is bounded and overwriting; backpressure to producers does not apply,
+//     except that producers can stall when wrapping to a slot currently being read.
+package broadcast
+
+import (
+	"math/bits"
+	"sync"
+	"sync/atomic"
+)
+
+// EventOrLag is a tagged union:
+//   - If isLag is false, event holds a regular message.
+//   - If isLag is true, lag holds the lag notification.
+type EventOrLag[T any] struct {
+	event T           // For a regular message
+	lag   LaggedError // For a lag notification
+	isLag bool
+}
+
+// IsEvent returns true if the EventOrLag contains a regular Event.
+func (e *EventOrLag[T]) IsEvent() bool {
+	return !e.isLag
+}
+
+// IsLag returns true if the EventOrLag contains a lag notification.
+func (e *EventOrLag[T]) IsLag() bool {
+	return e.isLag
+}
+
+// Event returns the event value or an error if none.
+func (e *EventOrLag[T]) Event() (T, error) {
+	var zero T
+	if e.isLag {
+		return zero, ErrNoEvent
+	}
+	return e.event, nil
+}
+
+// Lag returns the lag info or an error if none.
+func (e *EventOrLag[T]) Lag() (LaggedError, error) {
+	if !e.isLag {
+		return LaggedError{}, ErrNoLag
+	}
+	return e.lag, nil
+}
+
+// Subscription represents a single consumer.
+// Fields:
+// - bcast: back reference to broadcast
+// - nextSeq: next sequence number this subscriber expects to receive.
+// - out:     user-facing channel delivering EventOrLag values.
+// - done:    internal stop signal for this subscription.
+//
+// Lifecycle:
+//   - Created by Broadcast.Subscribe, which spawns a delivery goroutine (run).
+//   - Call Unsubscribe to stop; it requests the goroutine to exit promptly and
+//     wakes it if blocked, then the goroutine closes out upon return.
+//   - If Broadcast.Close is called, subscriptions will only exit once
+//     they can observe the closed context in their run loop, after draining the channel.
+type Subscription[T any] struct {
+	bcast     *Broadcast[T]
+	nextSeq   atomic.Uint64
```

Comment:
Can revert this back to `uint64`? I think it's only used in the `run` loop, except in cleanup code. If we do the cleanup properly, then I think we can void having another atomic value here.